Introduce socket service API

jukkar opened the pull request:
This PR introduces a socket service API. It can be used to create one socket listener that calls a callback if a set of sockets have any activity. This saves memory because instead of creating multiple threads for a set of sockets, one main thread can listen all needed sockets. This is similar concept how inetd works in Linux, although it is implemented differently.

jukkar committed:
net: sockets: Create a socket service API

The socket service provides a similar functionality as what
initd provides in Linux. It listens user registered sockets
for any activity and then launches a k_work for it. This way
each application does not need to create a thread to listen
a blocking socket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: service: Add tests for socket service API

Simple tests that verify that the socket service API works
as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: sockets: Add echo-service sample

The echo-service sample demostrates how to use the socket
service API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
This is still a draft, we can change the APIs etc, or discard the PR if it is not needed. Anyway, please take a look.

rlubos commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,245 @@
+		goto fail;
+	}
+
+	/* Create contiguous k_poll_event array to enable socket polling */
```
Is this comment in the right place? I don't see releavance to `k_poll_event` here.

jukkar replied:
Yes, it was a leftover from previous version where I was planning to use `k_poll_event` directly, but that did not work as planned.

rlubos commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,245 @@
+	}
+
+	/* The +1 is for triggering events from register function */
+	ctx.events = k_malloc(sizeof(struct zsock_pollfd) * (count + 1));
```
I wonder if we could avoid the heap dependency. Zephyr's `poll()` already has upper limit for monitored entries (`NET_SOCKETS_POLL_MAX `) so maybe we could just use it? And report an error if we exceed the limit.

pdgendt replied:
We could make a separate named linker section with the `struct zsock_pollfd` instances for the services.

jukkar replied:
I will investigate this. I would like to get rid of malloc too.

rlubos commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,180 @@
+ * @param cb User-supplied callback function to call
+ * @param user_data User specified data
+ */
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data);
```
I wonder, what's intended use case for such functionality?

jukkar replied:
It was for the net-shell. I was planning to have some monitoring information provided by shell and there this function would be useful. I did not implemented the shell part yet as so many things are still open with the APIs etc.

pdgendt commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -73,6 +73,31 @@ config NET_SOCKET_MAX_SEND_WAIT
+config NET_SOCKETS_SERVICE
+	bool "Enable sockets service support [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	select EVENTFD
```
I've had issues in the past with `EVENTFD` on some boards (I think it was `native_posix`), for that reason I've used `NET_SOCKETPAIR` in the CoAP server implementation.

Not sure if this is still an issue though.

jukkar replied:
There was some issues with `native_sim`, I will investigate this more.

jukkar replied:
Hmm, in eventfd Kconfig there is `depends on !NATIVE_APPLICATION` and the new `echo-service` process does not properly handle the incoming traffic with this board. I would like to avoid adding new dependency to socketpair just for this.

rlubos commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,180 @@
+ * User should create needed sockets and then setup the poll struct and
+ * then register the sockets to be monitored at runtime.
+ */
+struct net_socket_service_desc {
```
I wonder if we really need this static configuration for socket service? It seems a bit problematic, at least for my use case - DHCP server. Here, ideally, I'd need to have a single socket for each interface that wants to support the DHCP server functionality. But I don't think we have a way to calculate the number of interfaces in the system at build time to supply the static configuration macros. Purely runtime registration would be less cumbersome in such case.

jukkar replied:
> I wonder if we really need this static configuration for socket service? It seems a bit problematic, at least for my use case - DHCP server. Here, ideally, I'd need to have a single socket for each interface that wants to support the DHCP server functionality. But I don't think we have a way to calculate the number of interfaces in the system at build time to supply the static configuration macros. Purely runtime registration would be less cumbersome in such case.

I originally did some experimentation with runtime configuration, but as the API requires some housekeeping activities, allocating them at runtime requires malloc etc. There is a call to malloc at this PR too but hopefully we could get rid of it.

We could place `struct zsock_pollfd` into DHCP config in network interface, that way the information this API needs, would already be part of network interface.

cfriedt replied:
Devicetree could help here, just sayin 🤷🏼‍♂️

rlubos commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,245 @@
+		return -ENOENT;
+	}
+
+	if (svc->work_q != NULL) {
```
Shouldn't we call the registred callback directly from the services thread? I see a potential race here - we do submit the work to execute, but don't read the data (in case POLLIN was monitored) from the socket until the work executes. If the workqueue thread has lower prio than the services thread, we may end up busy looping  (poll() will keep reporting POLLIN as the work item didn't have a chance to execute).

jukkar replied:
> Shouldn't we call the registred callback directly from the services thread?

I originally planned to do that, but it felt more natural to use the existing APIs and resources in the system. So using a work queue does the same thing. But we can certainly change this, it is just to change one function call.

When using workqueue(s), we can serve (at least theoretically) multiple socket services at the "same" time and the priority of the thread using the services api determines which one gets called first. If we have a callback directly from services thread, then there is no such thing available.

jukkar replied:
> I see a potential race here - we do submit the work to execute, but don't read the data (in case POLLIN was monitored) from the socket until the work executes.

Note that `k_work_submit()` does nothing if the work is already pending. It is indeed possible that we might busyloop in the service thread if the service user does not "run" fast enough. I tried to mitigate this problem by setting the service thread priority to lowest application thread priority.

jukkar replied:
There seems to an issue when calling the "callback" via the workqueue as it might be possible (depending on different thread priorities) that we start to loop through poll() if the callback does not get run properly. So it might be better if we call the callback directly (synchronously) without using the workqueue. I will investigate this more.

jukkar replied:
> There seems to an issue when calling the "callback" via the workqueue as it might be possible (depending on different thread priorities) that we start to loop through poll() if the callback does not get run properly. So it might be better if we call the callback directly (synchronously) without using the workqueue. I will investigate this more.

Left the async call but fixed the code so that the callback work is only called once. Fixed also the tests that were not working properly with native_sim.

rlubos replied:
To be honest I'm not personally convinced about this async approach. Having multiple work queues kind of busts the idea the socket services were created for (saving resources), so most of the handlers will likely end up on a system workqueue anyway. Plus, having the handlers to be called directly from the services thread allows to reuse the services stack and kind of finetune its size for the worst case scenario (I imagine all of the registered services would need to allocate buffers for socket send/recv calls and a common stack area seemed like a good place for this instead of using static/heap/custom stack memory).

But it's just my personal opinion, no strong push from my side to change it, if that's the preferred way.

jukkar replied:
Personally I prefer the async way of doing things although it certainly complicates things. But I can see that sync way is also possible here. I think we can support both methods quite easily and let the user select how the callback should be called. I will propose a new version that supports both methods.

cfriedt replied:
Yes - sync should be a relatively trivial wrapper around async ... or at least I thought it was... 🤔

cfriedt commented:
Looks reasonable. It would be great to get a compile-time const number for network interfaces. Devicetree can fix that 😅

jukkar commented:
> It would be great to get a compile-time const number for network interfaces. Devicetree can fix that

Indeed, that info would help a lot if various part of the network stack. Although this PR should not depend on number of network interfaces.

jukkar commented:
Updates:
* removed malloc and allocated static array of pollfd in the size of `CONFIG_NET_SOCKETS_POLL_MAX`
* tweak the tests so that they are not run in `native_sim` board (because eventfd does not work there)

cfriedt commented:
> * tweak the tests so that they are not run in `native_sim` board (because eventfd does not work there)

Ugh.. that's annoying. I kept getting reassurances that posix arch and posix api were not mutually exclusive, but I often see that is still the case.

jukkar commented:
> Ugh.. that's annoying. I kept getting reassurances that posix arch and posix api were not mutually exclusive, but I often see that is still the case.

Indeed, that is bad. If I try to run the PR sample or the tests in native_sim/posix boards, they just "hang" and do not proceed. Pressing ctrl-c does not terminate the zephyr.exe even though the control is returned to user and the process looks like it terminated.

jukkar commented:
* Changed the sample to let CI pass.
* Rebased on top of latest main

jukkar committed:
tests: net: socket: Add correct path to socket.h for POSIX_API

If CONFIG_POSIX_API is enabled, then the socket.h is found under
zephyr/posix/sys/socket.h etc. This allows one to compile the
socket test applications without error prints.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* re-enabled native_sim after finding out why the test was hanging

jukkar commented:
> Ugh.. that's annoying. I kept getting reassurances that posix arch and posix api were not mutually exclusive, but I often see that is still the case.

With help from Alberto managed to get the eventfd working in the test  app. All I needed was to set `CONFIG_POSIX_API` and making sure picolibc was used. The hang issue was caused by an error in my test application and it was not related with eventfd or native_sim.

cfriedt commented:
> With help from Alberto managed to get the eventfd working in the test  app. All I needed was to set `CONFIG_POSIX_API` and making sure picolibc was used. The hang issue was caused by an error in my test application and it was not related with eventfd or native_sim.

Awesome - the libc selection should be completely irrelevant. Was there something in particular that was available in picolibc that was not available in the minimal libc that fixed your issue?

> @cfriedt Just wondering if we should select `CONFIG_POSIX_API` automatically or have eventfd depend on it so that user does not need to select it manually, or similar.

That would be one option. The only downside to that is that it might unnecessarily activate a bunch of unnecessary things.

> To be honest, I am a bit lost with various posix options and what needs to done with different combinations.

Simplifying configuration is on the roadmap. So is switching to standard option names, so that there is at least a direct mapping to the specification.

I would like to have 3 shortcut Kconfig options for PSE51, PSE52, and PSE53, as well as additional shortcut Kconfig options for subprofiling options and option groups.

Could I ask you to review a PR for that?

jukkar commented:
* fixed CI

cfriedt reviewed (dismissed):
Looks ok to me. I like that this is "threadless" and I think I know where this is going.

jukkar commented:
Added both sync and async way of calling the callback. The sample is modified to show how that can be done.

jukkar commented on samples/net/sockets/echo_service/sample.yaml:
```diff
@@ -0,0 +1,18 @@
+  filter: CONFIG_FULL_LIBC_SUPPORTED and not CONFIG_NATIVE_LIBC
+  # eventfd does not work properly with native_sim so exclude it here
+  platform_exclude:
+    - native_sim
+    - native_sim_64
```
Note to self. Forgot to fix this, the sample app works just fine with native_sim so I will fix this.

jukkar commented on samples/net/sockets/echo_service/src/main.c:
```diff
@@ -0,0 +1,254 @@
+			continue;
+		}
+
+		printk("----------client %d\n", client);
```
Missed this, will fix.

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -73,6 +73,31 @@ config NET_SOCKET_MAX_SEND_WAIT
+
+config NET_SOCKETS_SERVICE_STACK_SIZE
+	int "Stack size for the thread handling socket services"
+	default 1024
```
Note to self, this needs to be increased to 1200 as the usage was over 90% in my tests.

jukkar commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,301 @@
+			 "%d poll entries configured.",
+			 count, ARRAY_SIZE(ctx.events) - 1);
+		NET_WARN("Consider increasing value of %s to %d",
+			 "CONFIG_NET_SOCKETS_POLL_MAX", count);
```
Note to self, the eventfd was missing so the poll max needs to be at least `count + 1`

jukkar committed:
net: shell: Add sockets services prints

The socket services users to "net sockets" command.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Added service information printout to net-shell `net sockets` command.

pdgendt commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,246 @@
+#define NET_SOCKET_SERVICE_OWNER
+#endif
+
+#define NET_SOCKET_SERVICE_CALLBACK_MODE(_flag)				\
+	COND_CODE_1(_flag,						\
+		    (.work = Z_WORK_INITIALIZER(net_socket_service_callback),), ())
```
```suggestion
#define NET_SOCKET_SERVICE_CALLBACK_MODE(_flag)				\
	IF_ENABLED(_flag,						\
		    (.work = Z_WORK_INITIALIZER(net_socket_service_callback),))
```

pdgendt commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,301 @@
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
```
You already have the pointer, why still look it up? You might verify that the pointer is inside the linker section though.

pdgendt commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,301 @@
+
+static void service_callback_sync(struct k_work *work)
+{
+	struct net_socket_service_event *pev =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct net_socket_service_event ev = *pev;
+
+	ev.callback(&ev.work);
```
What is happening here? Are you copying the event?
If so, why? And does this work with nested `work` objects?

jukkar replied:
I did not want the callback to modify the event.
What nested work objects are you referring to?

pdgendt replied:
`ev.work`, or isn't that a `struct k_work` instance?

jukkar replied:
Sure, but what kind of issue you see there. This work is not embedded to anything user has supplied and the API does not access the copied work.

jukkar commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,246 @@
+};
+
+/**
+ * Socket service is a struct that user needs to fill properly.
```
I changed the first sentence to `Main structure holding socket service configuration information.` as the state information is not stored in this struct (because it is stored in rom).

jukkar commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,246 @@
+ * User should create needed sockets and then setup the poll struct and
+ * then register the sockets to be monitored at runtime.
+ */
+struct net_socket_service_desc {
```
The name could be that too. A similar concept was used in the still pending HTTP server PR so I followed suit here. I think I will use the `_desc` postfix here to make it clear there are two structs involved, a descriptor and then the event, both of which have a postfix attached to them.

jukkar commented on include/zephyr/net/socket_service.h:
```diff
@@ -0,0 +1,246 @@
+ *
+ * The socket service can be accessed outside the module where it is defined using:
+ *
+ * @code extern struct net_socket_service <name>; @endcode
```
Will fix this one.

kartben replied:
Sure, works for me. Mind you, there's two occurrences to replace in the PR.
Thanks!

jukkar committed:
net: shell: Avoid gcc warning print with string catenation

gcc prints this warning message

'strncat' specified bound 1 equals source length [-Wstringop-overflow=]
   58 |                 strncat(fd, "C", 1);

There was no error in the code but avoid the warning by not using
strncat().

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Updated according to comments
* Changed the code to use only one callback function. Last week I had some issues there so made separate callbacks for sync and async version. Turns out that I had a bugs when handling the global event array and fixing those let the api to use one callback func.

kartben reviewed (dismissed):
Looks good from docs standpoint, thanks!

pdgendt commented on samples/net/sockets/echo_service/prj.conf:
```diff
@@ -0,0 +1,38 @@
+
+# Network address config
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_NEED_IPV4=y
+CONFIG_NET_CONFIG_NEED_IPV6=y
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"
```
I only see IPv6 being used in the sample?

jukkar replied:
The sample supports both IPv4 and IPv6 by turning off the V6ONLY socket option.

pdgendt commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,286 @@
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	if (STRUCT_SECTION_START(net_socket_service_desc) >= svc &&
+	    STRUCT_SECTION_END(net_socket_service_desc) < svc) {
+		goto out;
+	}
```
This doesn't seem right.

```suggestion
	if (STRUCT_SECTION_START(net_socket_service_desc) > svc ||
	    STRUCT_SECTION_END(net_socket_service_desc) <= svc) {
		goto out;
	}
```

jukkar replied:
Indeed, will fix.

rlubos commented on subsys/net/lib/sockets/sockets_service.c:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	event->svc = svc;
+
```
Tried this with DHCPv4 server service, generally seems to work, however one thing brought my attention.

In the callback function registered for the service I was checking `event.revents` field of the `struct net_socket_service_event` provided to the callback (to check if there's no POLLERR on a socket, or whether it's really POLLIN that triggered the cb). It seems however that `event->event.revents` is not set here, so the callback could not inspect `revents` as it was always 0.

To me it looks simply like an overlook, but perhaps I miss something, is that intended behavior?

jukkar replied:
> To me it looks simply like an overlook, but perhaps I miss something, is that intended behavior?

No, that was not intended. Will fix.

rlubos reviewed (changes requested):
One issue spotted when testing.

jukkar commented:
New version fixes the issue that Robert was seeing.

rlubos reviewed (approved):
Thanks
